File: psana_examples/tags/V00-00-13/src/DumpAcqiris.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpAcqiris...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpAcqiris.h"

//-----------------
// C/C++ Headers --
//-----------------
#include <algorithm>

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "pdsdata/xtc/DetInfo.hh"
#include "psddl_psana/acqiris.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpAcqiris)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpAcqiris::DumpAcqiris (const std::string& name)
  : Module(name)
{
}

//--------------
// Destructor --
//--------------
DumpAcqiris::~DumpAcqiris ()
{
}

/// Method which is called at the beginning of the calibration cycle
void 
DumpAcqiris::beginCalibCycle(Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  Source src = configStr("source", "DetInfo(:Acqiris)");
  shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(src, &m_src);
  if (acqConfig.get()) {
    WithMsgLog(name(), info, str) {
      str << "Acqiris::ConfigV1: nbrBanks=" << acqConfig->nbrBanks()
          << " channelMask=" << acqConfig->channelMask()
          << " nbrChannels=" << acqConfig->nbrChannels()
          << " nbrConvertersPerChannel=" << acqConfig->nbrConvertersPerChannel();
     
      const Psana::Acqiris::HorizV1& h = acqConfig->horiz();
      str << "\n  horiz: sampInterval=" << h.sampInterval()
           << " delayTime=" << h.delayTime()
           << " nbrSegments=" << h.nbrSegments()
           << " nbrSamples=" << h.nbrSamples();
      
      for (unsigned ch = 0; ch < acqConfig->nbrChannels() ; ++ ch) {
        const Psana::Acqiris::VertV1& v = acqConfig->vert(ch);
        str << "\n  vert(" << ch << "):"
            << " fullScale=" << v.fullScale()
            << " slope=" << v.slope()
            << " offset=" << v.offset()
            << " coupling=" << v.coupling()
            << " bandwidth=" << v.bandwidth();
      }
    }
  }
}

/// Method which is called with event data, this is the only required 
/// method, all other methods are optional
void 
DumpAcqiris::event(Event& evt, Env& env)
{

  Pds::Src src;
  shared_ptr<Psana::Acqiris::DataDescV1> acqData = evt.get(m_src);
  if (acqData.get()) {
    
    // find matching config object
    shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(m_src);
    
    // loop over channels
    int nchan = acqData->data_shape()[0];
    for (int chan = 0; chan < nchan; ++ chan) {
      
      const Psana::Acqiris::DataDescV1Elem& elem = acqData->data(chan);

      const Psana::Acqiris::VertV1& v = acqConfig->vert(chan);
      double slope = v.slope();
      double offset = v.offset();

      WithMsgLog(name(), info, str ) {

        str << "Acqiris::DataDescV1: channel=" << chan
           << "\n  nbrSegments=" << elem.nbrSegments()
           << "\n  nbrSamplesInSeg=" << elem.nbrSamplesInSeg()
           << "\n  indexFirstPoint=" << elem.indexFirstPoint();
        
        // loop over segments
        for (unsigned seg = 0; seg < elem.nbrSegments(); ++ seg) {
          
          str << "\n  Segment #" << seg
              << "\n    timestamp=" << elem.timestamp(seg).pos()
              << "\n    data=[";
          
          const int16_t* wf = elem.waveforms() + seg*elem.nbrSamplesInSeg();

          unsigned size = std::min(elem.nbrSamplesInSeg(), 32U);
          for (unsigned i = 0; i < size; ++ i) {
            str << (wf[i]*slope + offset) << ", ";
          }
          str << "...]";
        
        }
      }
    }
  }
}

} // namespace psana_examples

File: psana_examples/trunk/src/DumpAcqiris.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_simple.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Class DumpAcqiris...
#
# Author List:
#  Joseph S. Barrera III
#------------------------------------------------------------------------

"""Example module for accessing SharedIpimb data.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id: dump_simple.py 2622 2011-11-11 14:35:00Z salnikov@SLAC.STANFORD.EDU $

@author Joseph S. Barrera III
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision: 2622 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

# local definitions usually start with _

#---------------------
#  Class definition --
#---------------------
class DumpAcqiris (object) :
    """Class whose instance will be used as a user analysis module. """

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        pass

    #-------------------
    #  Public methods --
    #-------------------
    def beginCalibCycle( self, evt, env ) :

        print "in beginCalibCycle()"

        # PSEvt::Source src = this->configStr("source", "DetInfo(:Acqiris)");
        # Source src = configStr("source", "DetInfo(:Acqiris)");
        src = env.configStr("source", "DetInfo(:Acqiris)")
        self.source = src # XXXXX

        # shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(src, &m_src);
        config = env.configStore().get("Psana::Acqiris::Config", src)
        print "---type(config)=", type(config)
        print "---config=", config

        if config:
        
            print "%s:" % (config.__class__.__name__)
            
            print "  nbrBanks =", config.nbrBanks(),
            print "channelMask =", config.channelMask(),
            print "nbrChannels =", config.nbrChannels(),
            print "nbrConvertersPerChannel =", config.nbrConvertersPerChannel()
     
            h = config.horiz()
            print "  horiz: sampInterval =", h.sampInterval(),
            print "delayTime =", h.delayTime(),
            print "nbrSegments =", h.nbrSegments(),
            print "nbrSamples =", h.nbrSamples()

            """
            try:
                vert = config.vert()
                print "vert=", vert
                nch = config.nbrChannels()
                for ch in range(nch):
                    v = config.vert2(ch)
                    print "  vert(%d):" % ch,
                    print "fullScale =", v.fullScale(),
                    print "slope =", v.slope(),
                    print "offset =", v.offset(),
                    print "coupling =", v.coupling(),
                    print "bandwidth=", v.bandwidth()
            except:
                print "vert.... oops"
            """

    def event(self, evt, env):
        acqData = evt.get("Psana::Acqiris::DataDesc", self.source)
        if acqData:

            # find matching config object
            acqConfig = env.configStore().get("Psana::Acqiris::Config", self.source)

            # loop over channels
            nchan = acqData.data_shape()[0];
            for chan in range(nchan):
                print "chan=", chan

                elem = acqData.data(chan);
                print "elem=", elem

                v_array = acqConfig.vert();
                """
                v = v_array.at(chan);
                print "v=", v
                slope = v.slope();
                print "slope=", slope
                offset = v.offset();
                print "offset=", offset
                """

                # WithMsgLog(name(), info, str ) {
                if True:

                    s = "Acqiris::DataDesc: channel=" + str(chan)
                    s = s + "\n  nbrSegments=" + str(elem.nbrSegments())
                    s = s + "\n  nbrSamplesInSeg=" + str(elem.nbrSamplesInSeg())
                    s = s + "\n  indexFirstPoint=" + str(elem.indexFirstPoint())
                    print s

        pass

    def endJob( self, evt, env ) :
        pass
